tests: flash: Fix disable_spi_nor filtering

decsny opened the pull request:
Require platform has a "storage partition" to run the "disable spi nor" case, due to this is what the test code expects.

Fixes #96682

Commit:
tests: flash: Fix disable_spi_nor case

This case was originally made sort of naively due to the fact that the
test was prioritizing testing spi nor first, but we wanted to test SOC
flash drivers, not the spi flash driver, so we just disabled
CONFIG_SPI_NOR. But this caused issues for platform which don't have a
storage partition or platform where the storage partition was actually
on a spi nor (because then there is no spi device leading to a linking
error with no reference to device object)

So fix by introducing a config to change the test logic to prioritize
the storage partition, which is really what we wanted to do in the first
place more directly I think.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

cfriedt commented:
@decsny - please check out the links in #96682 to learn which jobs are failing and how to verify that they are fixed.

Can you please also add a "Testing Done" section in the PR description to verify that this fixes the issues?

cfriedt commented:
```shell
west twister --build-only -c -p black_f407ve/stm32f407xx -p adafruit_feather_stm32f405/stm32f405xx -p gd32f450i_eval/gd32f450 -p gd32f450z_eval/gd32f450 -p legend@35/stm32f070xb -p mini_stm32h743/stm32h743xx -p pinetime_devkit0/nrf52832 -p rv32m1_vega/openisa_rv32m1/zero_riscy -T tests/drivers/flash/common -s drivers.flash.common.disable_spi_nor
Deleting output directory /Users/cfriedt/zephyrproject/zephyr/twister-out
INFO    - Using Ninja..
INFO    - Zephyr version: v4.2.0-4569-g20f149d5a805
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /Users/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
ERROR   - pinetime_devkit0/nrf52832 drivers.flash.common.disable_spi_nor               ERROR: Build failure
ERROR   - see: /Users/cfriedt/zephyrproject/zephyr/twister-out/pinetime_devkit0_nrf52832/zephyr/tests/drivers/flash/common/drivers.flash.common.disable_spi_nor/build.log
ERROR   - mini_stm32h743/stm32h743xx drivers.flash.common.disable_spi_nor              ERROR: Build failure
ERROR   - see: /Users/cfriedt/zephyrproject/zephyr/twister-out/mini_stm32h743_stm32h743xx/zephyr/tests/drivers/flash/common/drivers.flash.common.disable_spi_nor/build.log
INFO    - Total complete:    8/   8  100%  built (not run):    0, filtered:    6, failed:    0, error:    2
INFO    - 1 test scenarios (8 configurations) selected, 6 configurations filtered (0 by static filter, 6 at runtime).
INFO    - 0 of 2 executed test configurations passed (0.00%), 0 built (not run), 0 failed, 2 errored, with no warnings in 30.24 seconds.
INFO    - 0 test configurations executed on platforms, 2 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /Users/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /Users/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /Users/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) drivers.flash.common.disable_spi_nor on pinetime_devkit0/nrf52832 error (Build failure - undefined reference to `__device_dts_ord_95')
INFO    - 2) drivers.flash.common.disable_spi_nor on mini_stm32h743/stm32h743xx error (Build failure - undefined reference to `__device_dts_ord_139')
INFO    - 
INFO    - To rerun the tests, call twister using the following commandline:
INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
INFO    - 
INFO    - west twister -p mini_stm32h743/stm32h743xx -s drivers.flash.common.disable_spi_nor
INFO    - or with west:
INFO    - west build -p -b mini_stm32h743/stm32h743xx tests/drivers/flash/common -T drivers.flash.common.disable_spi_nor
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - Run completed
```

The error in the first log file is
```shell
/opt/zephyr/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `flash_erase':
/Users/cfriedt/zephyrproject/zephyr/twister-out/pinetime_devkit0_nrf52832/zephyr/tests/drivers/flash/common/drivers.flash.common.disable_spi_nor/zephyr/include/generated/zephyr/syscalls/flash.h:90: undefined reference to `__device_dts_ord_95'
/opt/zephyr/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `z_impl_flash_read':
/Users/cfriedt/zephyrproject/zephyr/include/zephyr/drivers/flash.h:274: undefined reference to `__device_dts_ord_95'
/opt/zephyr/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `flash_driver_before':
/Users/cfriedt/zephyrproject/zephyr/tests/drivers/flash/common/src/main.c:142: undefined reference to `__device_dts_ord_95'
/opt/zephyr/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `_flash_driver_test_flash_page_layout_wrapper':
/Users/cfriedt/zephyrproject/zephyr/tests/drivers/flash/common/src/main.c:362: undefined reference to `__device_dts_ord_95'
/opt/zephyr/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `flash_driver_test_read_unaligned_address':
/Users/cfriedt/zephyrproject/zephyr/tests/drivers/flash/common/src/main.c:201: undefined reference to `__device_dts_ord_95'
/opt/zephyr/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj):/Users/cfriedt/zephyrproject/zephyr/tests/drivers/flash/common/src/main.c:222: more undefined references to `__device_dts_ord_95' follow
```

decsny commented:
@cfriedt as I mentioned on the other ticket, I did realize what is the command used by weekly CI, and tried to replicate it but then realized there were unrelated build failures. The only point of trying to do that was in case any other baords than the one you mentioned were affected.

So I reverted to using your command in the issue (`west twister --build-only -c -p black_f407ve/stm32f407xx -p adafruit_feather_stm32f405/stm32f405xx -p gd32f450i_eval/gd32f450 -p gd32f450z_eval/gd32f450 -p legend@35/stm32f070xb -p mini_stm32h743/stm32h743xx -p pinetime_devkit0/nrf52832 -p rv32m1_vega/openisa_rv32m1/zero_riscy -T tests/drivers/flash/common -s drivers.flash.common.disable_spi_nor`) which only listed these 8 boards, and with the current state of PR, it did filter out 6 of the board but still tried to build 2 and errored. Then I realized what's the root issue is that this test is just written weirdly, and what we really directly want to do is not disable the spi nor but just test the storage partition with higher priority than the spi nor. So I fixed it in a different way, more directly, to how maybe the test case should have been in the first place. I will push in a minute an update which filtered the same 6 platform but built successfully the other 2

cfriedt reviewed (approved):
Awesome @decsny 🙌 Thanks for the fix!

decsny commented on tests/drivers/flash/common/testcase.yaml:
```diff
@@ -155,12 +155,12 @@ tests:
       - nrf5340dk/nrf5340/cpuapp
-  drivers.flash.common.disable_spi_nor:
-    filter: dt_compat_enabled("soc-nv-flash") and dt_compat_enabled("jedec,spi-nor")
+  drivers.flash.common.test_storage_partition:
```
No, because the point is to test the storage partition, I'm not fixing everyone other vendors lack of testing of their internal flash here, I'm fixing the CI issue by matching the test case to the test code

de-nordic replied:
The original idea behind the test, as far as I can tell, was to just test Flash API on flash for any vendor or mem type.
So the test required some place on flash to be able to play without destroying code.
It seems that we got lazy and decided that on external storage we can just weasel out of defining overlays with storage partition and select some arbitrary range.
Problem we have here is that the selector, on top of the test code, will prefer spi device if it is enabled; the code pr kinda fixes this.
The other option would be to remove the selector completely and rely on overlays providing the storage partition for each tested device - that would be much more work. Until we are ready to do this, I think what the PR does is fine.

JarmouniA reviewed (dismissed):
Won't block since this is meant to unblock CI, but this PR is simply removing the `force-test the internal flash when an external one is also available` test scenario, not fixing it, and adding an unrelated new one.

decsny commented:
> > Everything you said in comment is completely untrue. There was no "force test internal flash" case before, it was the "disable SPI nor because the test is hard coded to prioritize it over the storage partition" case, which is the default otherwise. And doing it that way it had the effect of breaking board which did in fact put the storage partition on the spi nor flash anyways. So you are completely wrong. This test CODE is not capable of testing an internal flash without the storage partition defined.
> 
> [d8be41c](https://github.com/zephyrproject-rtos/zephyr/commit/d8be41c00a1ca2dcdc97031436dc851a6f5137b9) :)
> 
> Also related #89924

Yes, thanks, that is my original commit explaining the exact same thing I am explaining now, and a PR from a vendor making an overlay to properly test their flash, exactly what they should be doing.

de-nordic commented:
> > > that is my original commit explaining the exact same thing I am explaining now
> > 
> > 
> > Really? From the commit message: _"add a twister case for disabling that config in case if a platform should also test some other flash which is present, such as internal SOC flash."_ How is this different from _"force-test the internal flash when an external one is also available"_.
> 
> Yes, the storage partition could be on a internal or external flash or whatever. That's the point.
> 
> But the way these lazy hacked in #ifdef were put by someone to test some spi nor instead of doing a little more effort of making an overlay, screwed up the test so that you couldn't test anything except the spi nor if the board had one enabled.
> 
> Please, read the actual code, and stop arguing things that are not true. If there is an internal flash with no storage partition defined, it indeed will not, but has never, and cannot be tested by this test in the current form, because the way this test is written relies on the storage partition being defined. So if you want to test some flash, you have always been supposed to define the storage partition on it.

https://github.com/zephyrproject-rtos/zephyr/pull/96749#discussion_r2409801098 basically what @decsny has stated here.
